src/System.Private.CoreLib/shared/System/MemoryExtensions.Trim.cs

```diff
@@ -0,0 +1,818 @@
+            where T : IEquatable<T>
+        {
+            // Use nested checks to avoid unnecessary branching for the unlikely case of N <= 1
+            if (trimElements.Length <= 1)
+            {
+                return trimElements.Length == 0 ? span : TrimEnd(span, trimElements[0]);
+            }
+
+            int length = ClampEnd(span, 0, trimElements);
+            return span.Slice(0, length);
+        }
+
+        /// <summary>
+        /// Delimits all leading occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        private static int ClampStart<T>(ReadOnlySpan<T> span, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            int start = 0;
+            for (; start < span.Length; start++)
+            {
+                if (!trimElements.Contains(span[start]))
+                {
+                    break;
+                }
+            }
+
+            return start;
+        }
+
+        /// <summary>
+        /// Delimits all trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="start">The start index from which to being searching.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        private static int ClampEnd<T>(ReadOnlySpan<T> span, int start, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            // Initially, start==len==0. If ClampStart trims all, start==len
+            Debug.Assert((uint)start <= span.Length);
+
+            int end = span.Length - 1;
+            for (; end >= start; end--)
+            {
+                if (!trimElements.Contains(span[end]))
+                {
+                    break;
+                }
+            }
+
+            return end - start + 1;
+        }
+
+        /// <summary>
+        /// Removes all leading and trailing white-space characters from the memory.
+        /// </summary>
+        public static Memory<char> Trim(this Memory<char> memory)
+        {
+            Span<char> span = memory.Span;
+
+            int start = ClampStart(span);
+            int length = ClampEnd(span, start);
+
+            return memory.Slice(start, length);
+        }
+
+        /// <summary>
+        /// Removes all leading white-space characters from the memory.
+        /// </summary>
+        public static Memory<char> TrimStart(this Memory<char> memory)
+            => memory.Slice(ClampStart(memory.Span));
+
+        /// <summary>
+        /// Removes all trailing white-space characters from the memory.
+        /// </summary>
+        public static Memory<char> TrimEnd(this Memory<char> memory)
```

Comment:
nit: For consistency, change to match `TrimEnd<T>`
```C#
        public static Memory<T> TrimEnd<T>(this Memory<T> memory, T trimElement)
            where T : IEquatable<T>
        {
            int length = ClampEnd(memory.Span, 0, trimElement);
            return memory.Slice(0, length);
        }
```

Reply:
Same, elsewhere TrimStart/etc.

Reply:
However, it's just a nit. Not worth resetting CI over.

Reply:
I don't mind, I can fix that nit.
Which idiom would you prefer; I can introduce the explaining var here, or I can change the other methods to inline the var and use expression bodies like this one?

Reply:
Fixed - I went with the more concise idiom; from previous feedback my understanding is that this would give it more of a chance of inlining since the var count is low.